apps/full-site-editing/full-site-editing-plugin/starter-page-templates/page-template-modal/index.js

```diff
@@ -29,46 +40,53 @@ class PageTemplateModal extends Component {
 	state = {
 		isLoading: false,
 		previewedTemplate: null,
-		blocksByTemplateSlug: {},
-		titlesByTemplateSlug: {},
 		error: null,
 		isOpen: false,
 	};
 
-	constructor( props ) {
-		super();
-		const hasTemplates = ! isEmpty( props.templates );
-		this.state.isOpen = hasTemplates;
-		if ( hasTemplates ) {
-			// Select the first template automatically.
-			this.state.previewedTemplate = this.getDefaultSelectedTemplate( props );
-			// Extract titles for faster lookup.
-			this.state.titlesByTemplateSlug = mapValues( keyBy( props.templates, 'slug' ), 'title' );
-		}
-	}
-
-	componentDidMount() {
-		if ( this.state.isOpen ) {
-			trackView( this.props.segment.id, this.props.vertical.id );
-		}
+	// Extract titles for faster lookup.
+	getTitlesByTemplateSlugs = memoize( templates =>
+		mapValues( keyBy( templates, 'slug' ), 'title' )
+	);
 
-		// Parse templates blocks and store them into the state.
-		const blocksByTemplateSlug = reduce(
-			this.props.templates,
+	// Parse templates blocks and store them into the state.
+	getBlocksByTemplateSlugs = memoize( templates =>
+		reduce(
+			templates,
 			( prev, { slug, content } ) => {
 				prev[ slug ] = content
 					? parseBlocks( replacePlaceholders( content, this.props.siteInformation ) )
 					: [];
 				return prev;
 			},
 			{}
-		);
+		)
+	);
+
+	static getDerivedStateFromProps( props, state ) {
+		// The only time `state.previewedTemplate` isn't set is before `templates`
+		// are loaded. As soon as we have our `templates`, we set it using
+		// `this.getDefaultSelectedTemplate`. Afterwards, the user can select a
+		// different template, but can never un-select it.
+		// This makes it a reliable indicator for whether the modal has just been launched.
+		if ( ! state.previewedTemplate && ! isEmpty( props.templates ) ) {
+			// Show the modal, and select the first template automatically.
+			return {
+				isOpen: true,
+				previewedTemplate: PageTemplateModal.getDefaultSelectedTemplate( props ),
+			};
+		}
+		return null;
+	}
 
-		// eslint-disable-next-line react/no-did-mount-set-state
-		this.setState( { blocksByTemplateSlug } );
+	componentDidUpdate( prevProps, prevState ) {
+		// Only track when the modal is first displayed.
+		if ( ! prevState.isOpen && this.state.isOpen ) {
```

Comment:
This doesn't get triggered when templates are passed directly from `window` object (current FSE context). Because the way `getDerivedStateFromProps` works, after first component update, `prevState.isOpen` is already `true`.

Reply:
Since this doesn't get called in the current FSE context, it may be worth leaving the `componentDidMount` check in place as well for now.  (unless we plan to hold off merging this until other changes happen?)

```
	componentDidMount() {
		if ( this.state.isOpen ) {
			trackView( this.props.segment.id, this.props.vertical.id );
		}
```
So if that is true on cDM, `trackView` will be called here and not in the cDU.  And if it becomes the case where that is not true on cDM, it would eventually get called in the new cDU.

Reply:
Good point, thanks @razvanpapadopol and @Addison-Stavlo!

I think that strategy that you suggest makes sense, @Addison-Stavlo. I'll add cDM back.